### PR TITLE
feat(id-austria): Add root patch and signature spoof patch

### PR DESCRIPTION
**Pull request by 1fexd**
Removes the root and bootloader open check from [ID Austria](https://play.google.com/store/apps/details?id=at.gv.oe.app&hl=de_AT&gl=US). It also spoofs the correct signature since the app checks the signature of the installed package.

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/idaustria/fingerprints/RootDetectionFingerprint.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+import app.revanced.patcher.fingerprint.method.impl.MethodFingerprint
+import org.jf.dexlib2.AccessFlags
+
+object RootDetectionFingerprint : MethodFingerprint(
+    "V",
+    access = AccessFlags.PUBLIC.value,
+    customFingerprint = { methodDef ->
+        methodDef.definingClass.endsWith("/DeviceIntegrityCheck;")
+                && (methodDef.name == "checkIntegrity" || methodDef.name == "checkIntegrityForceCheck")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/fingerprints/RootDetectionFingerprint.kt:11]: This check is redundant.

File: src/main/kotlin/app/revanced/patches/idaustria/fingerprints/SpoofSignatureFingerprint.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package app.revanced.patches.idaustria.fingerprints
+
+import app.revanced.patcher.fingerprint.method.impl.MethodFingerprint
+import org.jf.dexlib2.AccessFlags
+
+object SpoofSignatureFingerprint : MethodFingerprint(
+    "Ljava/lang/String",
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/fingerprints/SpoofSignatureFingerprint.kt:7]: The full qualified name is not checked, only the first character. Same goes for any other descriptor.

File: src/main/kotlin/app/revanced/patches/idaustria/patch/RootDetectionPatch.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import app.revanced.patcher.data.BytecodeContext
+import app.revanced.patcher.extensions.addInstructions
+import app.revanced.patcher.patch.BytecodePatch
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.idaustria.fingerprints.RootDetectionFingerprint
+
+@Patch
+@Name("root-patch")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/patch/RootDetectionPatch.kt:13]: Apply this change to the class name as well.
```suggestion
@Name("remove-root-detection")
```

File: src/main/kotlin/app/revanced/patches/idaustria/patch/RootDetectionPatch.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+@Description("Remove the root and bootloader check")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/patch/RootDetectionPatch.kt:14]:
```suggestion
@Description("Removes the check for root permissions and unlocked bootloader.")
```

File: src/main/kotlin/app/revanced/patches/idaustria/patch/RootDetectionPatch.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package app.revanced.patches.idaustria.patch
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/patch/RootDetectionPatch.kt:1]: Please move this patch and other related code to a proper package such as `app.revanced.patches.idaustria.detection.root`

File: src/main/kotlin/app/revanced/patches/idaustria/patch/SpoofSignaturePatch.kt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import app.revanced.patcher.extensions.addInstructions
+import app.revanced.patcher.patch.BytecodePatch
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.idaustria.fingerprints.SpoofSignatureFingerprint
+
+@Patch
+@Name("spoof-signature")
+@Description("Spoof the correct app-fingerprint")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/patch/SpoofSignaturePatch.kt:14]: Apply this change to the class name as well.

```suggestion
@Name("signature-spoof")
@Description("Spoofs the signature of the app.")
```

File: src/main/kotlin/app/revanced/patches/idaustria/patch/SpoofSignaturePatch.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
+package app.revanced.patches.idaustria.patch
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/patch/SpoofSignaturePatch.kt:1]: Please move this patch and other related code to a proper package such as `app.revanced.patches.idaustria.detection.signature`

File: src/main/kotlin/app/revanced/patches/idaustria/patch/SpoofSignaturePatch.kt [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.idaustria.fingerprints.SpoofSignatureFingerprint
+
+@Patch
+@Name("spoof-signature")
+@Description("Spoof the correct app-fingerprint")
+@Compatibility([Package("at.gv.oe.app", arrayOf("2.5.2"))])
+@Version("0.0.1")
+class SpoofSignaturePatch : BytecodePatch(listOf(SpoofSignatureFingerprint)) {
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/patch/SpoofSignaturePatch.kt:17]: Please fix formatting in this class.

File: src/main/kotlin/app/revanced/patches/idaustria/patch/SpoofSignaturePatch.kt [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    companion object {
+        const val signature =
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/patch/SpoofSignaturePatch.kt:19]: Constants have to be written in big letters: `EXPECTED_SIGNATURE`

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/idaustria/fingerprints/RootDetectionFingerprint.kt (hunk 1 above)
Comment (1fexd) [reply src/main/kotlin/app/revanced/patches/idaustria/fingerprints/RootDetectionFingerprint.kt:11]: Which check exactly? The method name check?

File: src/main/kotlin/app/revanced/patches/idaustria/detection/signature/patch/SpoofSignaturePatch.kt [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    override fun execute(context: BytecodeContext): PatchResult {
+        SpoofSignatureFingerprint.result!!.mutableMethod.addInstructions(
+            0,
+            """
+                   const-string v0, "$EXPECTED_SIGNATURE"
+                   return-object v0 
+                """
```
Comment (FineFindus) [inline src/main/kotlin/app/revanced/patches/idaustria/detection/signature/patch/SpoofSignaturePatch.kt:31]: Please fix the formatting.

Comment (FineFindus) [reply src/main/kotlin/app/revanced/patches/idaustria/detection/signature/patch/SpoofSignaturePatch.kt:31]: The `"""` should have the same indentation. Currently, the second one is a lot further indented.

Comment (1fexd) [reply src/main/kotlin/app/revanced/patches/idaustria/detection/signature/patch/SpoofSignaturePatch.kt]: Fixed

File: src/main/kotlin/app/revanced/patches/idaustria/detection/signature/patch/SpoofSignaturePatch.kt [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    listOf(SpoofSignatureFingerprint)
+) {
+    companion object {
+        const val EXPECTED_SIGNATURE =
+            "OpenSSLRSAPublicKey{modulus=ac3e6fd6050aa7e0d6010ae58190404cd89a56935b44f6fee067c149768320026e10b24799a1339e414605e448e3f264444a327b9ae292be2b62ad567dd1800dbed4a88f718a33dc6db6bf5178aa41aa0efff8a3409f5ca95dbfccd92c7b4298966df806ea7a0204a00f0e745f6d9f13bdf24f3df715d7b62c160090615de1c8a956b9286764985a3b3c060963c435fb9481a5543aaf0671fc2dba6c5c2b17d1ef1d85137f14dc9bbdf3490288087324cd48341cce64fabf6a9b55d1a7bf23b2fcdff451fd85bf0c7feb0a5e884d7c5c078e413149566a12a686e6efa70ae5161a0201307692834cda336c55157fef125e67c01c1359886f94742105596b42a790404bbcda5dad6a65f115aaff5e45ef3c28b2316ff6cef07aa49a45aa58c07bf258051b13ef449ccb37a3679afd5cfb9132f70bb9d931a937897544f90c3bcc80ed012e9f6ba020b8cdc23f8c29ac092b88f0e370ff9434e4f0f359e614ae0868dc526fa41e4b7596533e8d10279b66e923ecd9f0b200def55be2c1f6f9c72c92fb45d7e0a9ac571cb38f0a9a37bb33ea06f223fde8c7a92e8c47769e386f9799776e8f110c21df277ef1be61b2c01ebdabddcbf53cc4b6fd9a3c445606ee77b3758162c80ad8f8137b3c6864e92db904807dcb2be9d7717dd21bf42c121d620ddfb7914f7a95c713d9e1c1b7bdb4a03d618e40cf7e9e235c0b5687e03b7ab3,publicExponent=10001}"
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/detection/signature/patch/SpoofSignaturePatch.kt:22]:
```suggestion
        const val EXPECTED_SIGNATURE = "OpenSSLRSAPublicKey{modulus=ac3e6fd6050aa7e0d6010ae58190404cd89a56935b44f6fee067c149768320026e10b24799a1339e414605e448e3f264444a327b9ae292be2b62ad567dd1800dbed4a88f718a33dc6db6bf5178aa41aa0efff8a3409f5ca95dbfccd92c7b4298966df806ea7a0204a00f0e745f6d9f13bdf24f3df715d7b62c160090615de1c8a956b9286764985a3b3c060963c435fb9481a5543aaf0671fc2dba6c5c2b17d1ef1d85137f14dc9bbdf3490288087324cd48341cce64fabf6a9b55d1a7bf23b2fcdff451fd85bf0c7feb0a5e884d7c5c078e413149566a12a686e6efa70ae5161a0201307692834cda336c55157fef125e67c01c1359886f94742105596b42a790404bbcda5dad6a65f115aaff5e45ef3c28b2316ff6cef07aa49a45aa58c07bf258051b13ef449ccb37a3679afd5cfb9132f70bb9d931a937897544f90c3bcc80ed012e9f6ba020b8cdc23f8c29ac092b88f0e370ff9434e4f0f359e614ae0868dc526fa41e4b7596533e8d10279b66e923ecd9f0b200def55be2c1f6f9c72c92fb45d7e0a9ac571cb38f0a9a37bb33ea06f223fde8c7a92e8c47769e386f9799776e8f110c21df277ef1be61b2c01ebdabddcbf53cc4b6fd9a3c445606ee77b3758162c80ad8f8137b3c6864e92db904807dcb2be9d7717dd21bf42c121d620ddfb7914f7a95c713d9e1c1b7bdb4a03d618e40cf7e9e235c0b5687e03b7ab3,publicExponent=10001}"
```

File: src/main/kotlin/app/revanced/patches/idaustria/detection/signature/patch/SpoofSignaturePatch.kt [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import app.revanced.patcher.data.BytecodeContext
+import app.revanced.patcher.extensions.addInstructions
+import app.revanced.patcher.patch.BytecodePatch
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.idaustria.detection.signature.fingerprints.SpoofSignatureFingerprint
+
+@Patch
+@Name("signature-spoof")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/idaustria/detection/signature/patch/SpoofSignaturePatch.kt:13]:
```suggestion
@Name("spoof-signature")
```

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Review by oSumAtrIX — APPROVED** [review]

**Comment by oSumAtrIX** [comment]
Thanks.

**Pull request merged**